lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,917 @@
+                "ELSEWHERE statement name mismatch"_err_en_US});
+      }
+    }
+  }
+
+  // C1134, C1166
+  template<typename A>
+  void CheckLabelContext(const char *const stmtString, const A &constructName) {
+    const auto iter{std::find(
+        constructNames_.crbegin(), constructNames_.crend(), constructName)};
+    if (iter == constructNames_.crend()) {
+      errorHandler_.Say(currentPosition_,
+          parser::MessageFormattedText{
+              "%s construct-name '%s' is not in scope"_err_en_US, stmtString,
+              constructName.c_str()});
+    }
+  }
+
+  // 6.2.5, paragraph 2
+  void CheckLabelInRange(parser::Label label) {
+    if (label < 1 || label > 99999) {
+      errorHandler_.Say(currentPosition_,
+          parser::MessageFormattedText{
+              "label '%" PRIu64 "' is out of range"_err_en_US, label});
+    }
+  }
+
+  // 6.2.5., paragraph 2
+  void addTargetLabelDefinition(parser::Label label,
+      LabeledStmtClassificationSet labeledStmtClassificationSet) {
+    CheckLabelInRange(label);
+    const auto pair{programUnits_.back().targetStmts.emplace(label,
+        LabeledStatementInfoTuplePOD{
+            currentScope_, currentPosition_, labeledStmtClassificationSet})};
+    if (!pair.second) {
+      errorHandler_.Say(currentPosition_,
+          parser::MessageFormattedText{
+              "label '%" PRIu64 "' is not distinct"_err_en_US, label});
+    }
+  }
+
+  void addLabelReferenceFromDoStmt(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().doStmtSources.emplace_back(
+        label, currentScope_, currentPosition_);
+  }
+
+  void addLabelReferenceFromFormatStmt(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().formatStmtSources.emplace_back(
+        label, currentScope_, currentPosition_);
+  }
+
+  void addLabelReference(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().otherStmtSources.emplace_back(
+        label, currentScope_, currentPosition_);
+  }
+
+  void addLabelReference(const std::list<parser::Label> &labels) {
+    for (const parser::Label &label : labels) {
+      addLabelReference(label);
+    }
+  }
+
+  std::vector<UnitAnalysis> programUnits_;
+  parser::Messages errorHandler_;
+  parser::CharBlock currentPosition_{nullptr};
+  ProxyForScope currentScope_{0};
+  std::vector<std::string> constructNames_;
+};
+
+bool InInclusiveScope(const std::vector<ProxyForScope> &scopes,
+    ProxyForScope tail, ProxyForScope head) {
+  for (; tail != head; tail = scopes[tail]) {
+    if (!HasScope(tail)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+ParseTreeAnalyzer LabelAnalysis(const parser::Program &program) {
+  ParseTreeAnalyzer analysis;
+  Walk(program, analysis);
+  return analysis;
+}
+
+bool InBody(const parser::CharBlock &position,
+    const std::pair<parser::CharBlock, parser::CharBlock> &pair) {
+  if (position.begin() >= pair.first.begin()) {
```

Comment:
This code might be more clear if `CharBlock` and/or `Interval` were extended so that the endpoints of the pair could be combined into a single interval and then used for testing containment.